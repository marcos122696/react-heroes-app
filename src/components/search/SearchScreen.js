import React, { useMemo } from 'react';
import { useLocation, useNavigate } from 'react-router-dom';
import { useForm } from '../../hooks/useForm';
import { getHeroesByName } from '../../selectors/getHeroesByName';
import { HeroCard } from '../hero/HeroCard';
import queryString from 'query-string';

export const SearchScreen = () => {

    const navigate = useNavigate();
    const location = useLocation();

    const { q = '' } = queryString.parse(location.search);

    const [ formValues, handleInputChange ] = useForm({
        searchText: q
    });

    const { searchText } = formValues;
    const heroesFilter = useMemo( () => getHeroesByName( q ), [q] );

    const handleSearch = (e) => {
        e.preventDefault();
        navigate(`?q=${ searchText }`);
    };

  return (
    <>
        <h1 className="animate__animated animate__hinge">Busqueda</h1>
        <hr />

        <div className="row">
            <div className="col-5">
                <h4>Buscar Heroe</h4>
                <hr />

                <form onSubmit={ (e) => handleSearch(e) }>
                    <input 
                        type="text" 
                        placeholder="Nombre Heroe" 
                        className="form-control"
                        name="searchText"
                        autoComplete="off"
                        value={ searchText }
                        onChange={ handleInputChange }
                    />
                </form>

                <button
                    className="btn btn-outline-light mt-3 w-100"
                    type="subbmit"
                >
                    Buscar
                </button>
            </div>

            <div className="col-7">
                <h4>Resultados</h4>
                <hr />

                {
                        (q === '') 
                        ? <div className="alert alert-info">Busque un heroe</div>
                        : ( heroesFilter.length === 0) 
                            && <div className="alert alert-danger">No hay resultados de: {q}</div>
                }
                <ul className="d-flex flex-row flex-wrap">
                    {
                        heroesFilter.map( hero => (
                            <li className="card-item" key={ hero.id }>
                                <HeroCard
                                    { ...hero }
                                />   
                            </li>
                        ))
                    }
                </ul>
            </div>
        </div>
    </>
  )
}
