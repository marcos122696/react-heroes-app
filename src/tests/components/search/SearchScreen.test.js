import { mount } from "enzyme";
import { SearchScreen } from "../../../components/search/SearchScreen";
import { MemoryRouter } from "react-router-dom";


const mockNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockNavigate
}))

describe('Pruebas en el <SearchScreen />', () => {

    test('Debe de mostrarse correctamente con valores por defecto', () => {

        const wrapper = mount(
            <MemoryRouter initialEntries={ ['/search'] } >
                <SearchScreen />
            </MemoryRouter>
        );

        expect( wrapper ).toMatchSnapshot();
        expect( wrapper.find('input').prop('value') ).toBe('');
        expect( wrapper.find('.alert-info').text().trim() ).toBe('Busque un heroe');
    });

    test('Debe de mostrar a Batman y el input con el valor del queryString', () => {

        const wrapper = mount(
            <MemoryRouter initialEntries={ ['/search?q=batman'] } >
                <SearchScreen />
            </MemoryRouter>
        );

        expect( wrapper.find('input').prop('value') ).toBe('batman');
        expect( wrapper ).toMatchSnapshot();
    });

    test('Debe de mostarr un error si no se encuentra el heroe', () => {

        const wrapper = mount(
            <MemoryRouter initialEntries={ ['/search?q=heroe prueba'] } >
                <SearchScreen />
            </MemoryRouter>
        );

        
        expect( wrapper.find('.alert-danger').text().trim() ).toBe('No hay resultados de: heroe prueba')
    });

    test('Debe de llamar el navigate a la nueva pantalla', () => {

        const wrapper = mount(
            <MemoryRouter initialEntries={ ['/search'] } >
                <SearchScreen />
            </MemoryRouter>
        );

        wrapper.find('input').simulate('change', {
            target: {
                name: 'searchText',
                value: 'batman'
            }
        });
        wrapper.find('form').simulate( 'submit', {
            preventDefault: () => {},
        });

        expect( mockNavigate ).toHaveBeenCalledWith('?q=batman');
    });
});