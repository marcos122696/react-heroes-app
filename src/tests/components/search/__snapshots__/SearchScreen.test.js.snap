// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el <SearchScreen /> Debe de mostrar a Batman y el input con el valor del queryString 1`] = `
Array [
  <h1
    className="animate__animated animate__hinge"
  >
    Busqueda
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Buscar Heroe
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Nombre Heroe"
          type="text"
          value="batman"
        />
      </form>
      <button
        className="btn btn-outline-light mt-3 w-100"
        type="subbmit"
      >
        Buscar
      </button>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Resultados
      </h4>
      <hr />
      <ul
        className="d-flex flex-row flex-wrap"
      >
        <li
          className="card-item"
          key="dc-batman"
        >
          <div
            className="col"
          >
            <a
              href="/Hero/dc-batman"
              onClick={[Function]}
              style={
                Object {
                  "textDecoration": "none",
                }
              }
            >
              <div
                className="card bg-secondary text-white h-100"
              >
                <div
                  className="row no-gutters"
                >
                  <div
                    className="container-img"
                  >
                    <img
                      alt="Batman"
                      className="h-100 w-100 animate__animated animate__zoomIn"
                      src="/assets/dc-batman.jpg"
                    />
                  </div>
                  <div
                    className="col-8 w-100"
                  >
                    <div
                      className="card-body"
                    >
                      <h5
                        className="card-title"
                      >
                        Batman
                      </h5>
                      <p
                        className="card-text"
                      >
                        Bruce Wayne
                      </p>
                      <p
                        className="card-text"
                      >
                        <small
                          className="text-white-50"
                        >
                          Detective Comics #27
                        </small>
                      </p>
                    </div>
                  </div>
                </div>
              </div>
            </a>
          </div>
        </li>
      </ul>
    </div>
  </div>,
]
`;

exports[`Pruebas en el <SearchScreen /> Debe de mostrarse correctamente con valores por defecto 1`] = `
Array [
  <h1
    className="animate__animated animate__hinge"
  >
    Busqueda
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Buscar Heroe
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Nombre Heroe"
          type="text"
          value=""
        />
      </form>
      <button
        className="btn btn-outline-light mt-3 w-100"
        type="subbmit"
      >
        Buscar
      </button>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Resultados
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        Busque un heroe
      </div>
      <ul
        className="d-flex flex-row flex-wrap"
      />
    </div>
  </div>,
]
`;
